packages/lexicon-resolver/src/lexicon.ts

```diff
@@ -0,0 +1,120 @@
+import dns from 'node:dns/promises'
+import { CID } from 'multiformats/cid'
+import { LexiconDoc, parseLexiconDoc } from '@atproto/lexicon'
+import { Commit } from '@atproto/repo'
+import { AtUri, NSID, ensureValidDid } from '@atproto/syntax'
+import { ResolveRecordOptions, resolveRecord } from './record.js'
+import { isValidDid } from './util.js'
+
+const DNS_SUBDOMAIN = '_lexicon'
+const DNS_PREFIX = 'did='
+export const LEXICON_SCHEMA_NSID = 'com.atproto.lexicon.schema'
+
+export type ResolveLexiconOptions = ResolveRecordOptions & {
+  didAuthority?: string
+}
+
+export type LexiconResolution = {
+  commit: Commit
+  uri: AtUri
+  cid: CID
+  nsid: NSID
+  lexicon: LexiconDoc & LexiconSchemaRecord
+}
+
+/**
+ * Resolve a Lexicon from the network, verifying its authenticity.
+ * @param nsidStr NSID or string representing one for the Lexicon that will be resolved.
+ * @param options
+ * @returns
+ */
+export async function resolveLexicon(
+  nsidStr: NSID | string,
+  options: ResolveLexiconOptions = {},
+): Promise<LexiconResolution> {
+  const nsid = typeof nsidStr === 'string' ? NSID.parse(nsidStr) : nsidStr
+  const didAuthority = await getDidAuthority(nsid, options)
+  const verified = await resolveRecord(
+    AtUri.make(didAuthority, LEXICON_SCHEMA_NSID, nsid.toString()),
+    options,
+  ).catch((err) => {
+    throw new LexiconResolutionError(
+      'Could not resolve Lexicon schema record',
+      { cause: err },
+    )
+  })
+  let lexicon: LexiconDoc
+  try {
+    lexicon = parseLexiconDoc(verified.record)
+  } catch (err) {
+    throw new LexiconResolutionError('Invalid Lexicon document', { cause: err })
+  }
+  if (!isLexiconSchemaRecord(lexicon)) {
+    throw new LexiconResolutionError('Invalid Lexicon schema record')
+  }
+  if (lexicon.id !== nsid.toString()) {
+    throw new LexiconResolutionError(
+      `Lexicon schema record id does not match NSID: ${lexicon.id}`,
+    )
+  }
+  const { uri, cid, commit } = verified
+  return { commit, uri, cid, nsid, lexicon }
+}
+
+/**
+ * Resolve the DID authority for a Lexicon from the network using DNS, based on its NSID.
+ * @param nsidStr NSID or string representing one for which to lookup its Lexicon DID authority.
+ * @param options
+ * @returns
+ */
+export async function resolveLexiconDidAuthority(nsidStr: NSID | string) {
+  const nsid = typeof nsidStr === 'string' ? NSID.parse(nsidStr) : nsidStr
+  const did = await resolveDns(nsid.authority)
+  if (did == null || !isValidDid(did)) return
```

Comment:
nit: I think it's safer to explicit the return type of publicly exposed function to avoid any unexpected break in the future.